### PR TITLE
Implement timer for `getrandom()`

### LINKED ISSUE #507: Entropy problem on Nitro
**Describe the bug**
On some Nitro machines, requesting entropy by whether reading `/dev/random` or sending a `getrandom` syscall results in failure: the function call hangs and does not return immediately (or ever). This might be due to entropy shortage or incorrect entropy initialisation.
Passing `GRND_RANDOM` or `GRND_NONBLOCK` to `getrandom()` doesn't seem to make any difference.
Reading from `/dev/urandom` works.

**To Reproduce**
Call libc's (musl's) `getrandom()` or read `/dev/random` inside Nitro and wait for the return... forever (doesn't happen on every Nitro machine)

**Expected behaviour**
`getrandom()` or `/dev/random` should promptly return random bytes.

**Potential solutions**

##### 1. Use the NSM library
On Nitro, it is recommended to use AWS' NSM library for RNG.
The functions of interest are `nsm_get_random()` and `aws_nitro_enclaves_library_seed_entropy()`, the latter calling `nsm_get_random()` and seeding `/dev/random` with the returned random bytes.
It's curious that `/dev/random` is not seeded automatically by the kernel whenever the `getrandom` syscall is received.
Several approaches can be considered to redirect external dependencies to the right RNG:
* Use Veracruz's `platform-services`: In Veracruz, `platform-services` exposes `getrandom()` that calls functions from the NSM lib. WASM programs calling `getrandom()` are automatically redirected to it, but linking external dependencies, especially C code, against it is trickier (cf. #511)
* Mbed TLS specific: Pass a pointer to a custom RNG when initialising mbedtls (`mbedtls_ssl_conf_rng()`)

##### 2. Read RN from `/dev/urandom` (quick workaround)

### PR DESCRIPTION
**Pull request by gbryant-arm**
Since #507 hasn't been reproduced yet, we propose to print an error message whenever #507 shows up again. This shall be a signal to open a ticket on AWS.
TLDR: Detect but don't fix #507

### CODE REVIEW

**Comment by gbryant-arm** [comment]
Of course, icecap doesn't support threads... Replacing multithreading with signal handling

**Comment by gbryant-arm** [comment]
icecap doesn't support signal handling either?

**Comment by gbryant-arm** [comment]
Ready for review

**Comment by egrimley-arm** [comment]
Why is `entropy_poll.c` modified? Why can't everything be done in `mbedtls_hardware_poll.c`?

**Comment by egrimley-arm** [comment]
Previously, the contents of `mbedtls-sys/vendor` were identical to upstream `mbedtls-3.2.1`. Can we please keep it identical and just modify the wrapper rather than Mbed TLS itself?

**Comment by egrimley-arm** [comment]
Also, can we please keep the git history of `rust-mbedtls` nice and simple? It could be a single commit for this PR, surely?

**Comment by gbryant-arm** [comment]
> Why is `entropy_poll.c` modified? Why can't everything be done in `mbedtls_hardware_poll.c`?

I've simply greped every call to `getrandom()`. If you're certain this file is not used then I'll remove it.

> Previously, the contents of mbedtls-sys/vendor were identical to upstream mbedtls-3.2.1. Can we please keep it identical and just modify the wrapper rather than Mbed TLS itself?

I guess you're referring to `entropy_poll.c`. Cf. my answer above.

> Also, can we please keep the git history of rust-mbedtls nice and simple? It could be a single commit for this PR, surely?

Sure, let's pack the rust-mbedtls updates together

**Comment by egrimley-arm** [comment]
> If you're certain this file is not used then I'll remove it.

It's easy enough to check. The configuration is not dynamically generated, so if the code isn't used now, it won't be used in future.

**Comment by egrimley-arm** [comment]
We define `MBEDTLS_NO_PLATFORM_ENTROPY` so `mbedtls_platform_entropy_poll` does not get compiled on any of our platforms.

**Comment by egrimley-arm** [comment]
The diff looks good, but can we simplify the changes to `rust-mbedtls` to a single commit?

**Review by egrimley-arm — APPROVED** [review]

**Comment by mathias-arm** [comment]
The `gbryant-arm/getrandom-timer` branch in [rust-mbedtls](https://github.com/veracruz-project/rust-mbedtls) should have been merged into `veracruz` before this PR was merged. Opening veracruz-project/rust-mbedtls#6 (and closing veracruz-project/rust-mbedtls#3).
